File: app-android-biometric-passport-zk/app/src/main/java/org/iranUnchained/utils/PassportDataLoader.kt
package org.iranUnchained.utils

import android.content.Context
import android.util.Log
import com.google.gson.Gson
import com.google.gson.annotations.SerializedName
import org.iranUnchained.utils.nfc.model.EDocument
import org.iranUnchained.utils.nfc.model.PersonDetails
import java.io.File

/**
 * Loads passport data from a JSON file on the device filesystem.
 *
 * Expected path: /sdcard/Android/data/<applicationId>/files/passport-data.json
 * (accessible via context.getExternalFilesDir(null) — no runtime permissions needed)
 *
 * The JSON format matches the v1 export produced by the Android app's passport scan & export feature.
 */
object PassportDataLoader {
    private const val TAG = "PassportDataLoader"
    private const val FILENAME = "passport-data.json"

    data class PassportFilePersonDetails(
        val name: String? = null,
        val surname: String? = null,
        val nationality: String? = null,
        val issuerAuthority: String? = null,
        val dateOfBirth: String? = null,
        val dateOfExpiry: String? = null,
        val documentNumber: String? = null,
        val gender: String? = null
    )

    data class PassportFileData(
        val version: Int? = null,
        val type: String? = null,
        val exportedAt: String? = null,
        val dg1Hex: String? = null,
        val sodHex: String? = null,
        val digestAlgorithm: String? = null,
        val docSigningCertPem: String? = null,
        val personDetails: PassportFilePersonDetails? = null
    )

    /**
     * Attempts to load passport data from the device's app-specific external files directory.
     * Returns null if the file doesn't exist or can't be parsed.
     */
    fun loadFromDevice(context: Context): PassportFileData? {
        val dir = context.getExternalFilesDir(null) ?: run {
            Log.w(TAG, "External files dir not available")
            return null
        }
        val file = File(dir, FILENAME)
        if (!file.exists()) {
            Log.d(TAG, "No passport data file at ${file.absolutePath}")
            return null
        }

        return try {
            val json = file.readText()
            parseJson(json)?.also { data ->
                Log.i(TAG, "Loaded passport data: issuer=${data.personDetails?.issuerAuthority}, " +
                        "dob=${data.personDetails?.dateOfBirth}, " +
                        "name=${data.personDetails?.name} ${data.personDetails?.surname}")
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to read passport data file", e)
            null
        }
    }

    /**
     * Parses a passport JSON string. Pure function, testable on JVM.
     */
    fun parseJson(json: String): PassportFileData? {
        return try {
            Gson().fromJson(json, PassportFileData::class.java)
        } catch (e: Exception) {
            try { Log.e(TAG, "Failed to parse passport JSON", e) } catch (_: Exception) {}
            null
        }
    }

    /**
     * Builds an EDocument from loaded passport file data.
     * Used for Layer 2: full NFC bypass → ConfirmationActivity flow.
     */
    fun buildEDocument(data: PassportFileData): EDocument {
        val personDetails = data.personDetails?.let { pd ->
            PersonDetails(
                name = pd.name,
                surname = pd.surname,
                nationality = pd.nationality,
                issuerAuthority = pd.issuerAuthority,
                birthDate = pd.dateOfBirth,
                expiryDate = pd.dateOfExpiry,
                serialNumber = pd.documentNumber,
                gender = pd.gender
            )
        }

        return EDocument(
            personDetails = personDetails,
            sod = data.sodHex,
            dg1Hex = data.dg1Hex,
            digestAlgorithm = data.digestAlgorithm,
            docSigningCertPem = data.docSigningCertPem
        )
    }
}

File: app-ios-biometric-passport-zk/IranUnchained/Code/Utils/PassportDataLoader.swift
//
//  PassportDataLoader.swift
//  IranUnchained
//
//  Loads passport data from a JSON file for local development testing.
//  Port of Android's PassportDataLoader.kt.
//

import Foundation
import OSLog

enum PassportDataLoader {
    static let filename = "passport-data.json"

    struct PersonDetails: Codable {
        let name: String?
        let surname: String?
        let nationality: String?
        let issuerAuthority: String?
        let dateOfBirth: String?
        let dateOfExpiry: String?
        let documentNumber: String?
        let gender: String?
    }

    struct PassportFileData: Codable {
        let version: Int?
        let type: String?
        let exportedAt: String?
        let dg1Hex: String?
        let sodHex: String?
        let digestAlgorithm: String?
        let docSigningCertPem: String?
        let personDetails: PersonDetails?
    }

    /// Loads passport data from the app's Documents directory.
    /// Users can place the file via Finder, Files app, or `xcrun simctl` push.
    static func loadFromDocuments() -> PassportFileData? {
        guard let documentsDir = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first else {
            Logger.localDev.warning("Documents directory not available")
            return nil
        }

        let fileURL = documentsDir.appendingPathComponent(filename)

        guard FileManager.default.fileExists(atPath: fileURL.path) else {
            Logger.localDev.debug("No passport data file at \(fileURL.path)")
            return nil
        }

        do {
            let json = try String(contentsOf: fileURL, encoding: .utf8)
            let data = parseJson(json)
            if let data {
                Logger.localDev.info("Loaded passport data from Documents: issuer=\(data.personDetails?.issuerAuthority ?? "nil"), dob=\(data.personDetails?.dateOfBirth ?? "nil")")
            }
            return data
        } catch {
            Logger.localDev.error("Failed to read passport data file: \(error)")
            return nil
        }
    }

    /// Loads passport data from the app bundle (for simulator convenience).
    /// The file must be added to the Xcode project and included in the target.
    static func loadFromBundle() -> PassportFileData? {
        guard let fileURL = Bundle.main.url(forResource: "passport-data", withExtension: "json") else {
            Logger.localDev.debug("No passport-data.json in app bundle")
            return nil
        }

        do {
            let json = try String(contentsOf: fileURL, encoding: .utf8)
            let data = parseJson(json)
            if let data {
                Logger.localDev.info("Loaded passport data from bundle: issuer=\(data.personDetails?.issuerAuthority ?? "nil"), dob=\(data.personDetails?.dateOfBirth ?? "nil")")
            }
            return data
        } catch {
            Logger.localDev.error("Failed to read passport data from bundle: \(error)")
            return nil
        }
    }

    /// Parses a passport JSON string. Pure function, testable.
    static func parseJson(_ json: String) -> PassportFileData? {
        guard let jsonData = json.data(using: .utf8) else {
            return nil
        }

        do {
            return try JSONDecoder().decode(PassportFileData.self, from: jsonData)
        } catch {
            Logger.localDev.error("Failed to parse passport JSON: \(error)")
            return nil
        }
    }

    /// Attempts to load passport data from Documents first, then falls back to bundle.
    static func load() -> PassportFileData? {
        if let data = loadFromDocuments() {
            return data
        }
        return loadFromBundle()
    }
}
